llvm/include/llvm/CAS/ActionCache.h

```diff
@@ -0,0 +1,233 @@
+//===- llvm/CAS/ActionCache.h -----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CAS_ACTIONCACHE_H
+#define LLVM_CAS_ACTIONCACHE_H
+
+#include "llvm/CAS/ActionDescription.h"
+#include "llvm/Support/Error.h"
+
+namespace llvm {
+namespace cas {
+
+/// A cache of actions and results, mapping from an arbitrary StringRef to
+/// UniqueID.
+///
+/// Belongs to a specific \a Namespace. Must be in the same namespace as the
+/// IDs sent to \a put().
+class ActionCache {
+  virtual void anchor();
+
+public:
+  /// Put an action in the cache, such that \p Action maps to \p Result. Care
+  /// should be taken that the action always produces the same result by
+  /// including all inputs (e.g., a UniqueID for the clang binary's blob should
+  /// be included if caching something in the Clang compiler).
+  ///
+  /// Thread-safe.
+  ///
+  /// \return Error if there's a collision or some problem storing the result.
+  /// \pre \c Result.getNamespace() is the same as \a getNamespace().
+  /// \post Subsequent call to \a get() with \p Action returns \p Result.
+  /// Returns an error if there is a \a get() will not return
+  Error put(const ActionDescription &Action, UniqueIDRef Result) {
+    assert(&Result.getNamespace() == &getNamespace());
```

Comment:
The cache is decoupled from storage, but you need some way of knowing whether the UniqueIDs that they both refer to mean the same thing. Otherwise, the action cache might store an ID from an LLVM builtin CAS and then you try to look the object up in a custom CAS of some sort where the ID has a different meaning.

Sharing a "namespace" tells you that a UniqueID has the same meaning.

Reply:
I guess I wasn't clear what is covered by a Namespace. Is that Key, or Value, or the mapping between Key and Value?

The Value/Result from ActionCache should be available as Key in ObjectStore but is very different from Value/Result in ObjectStore. I guess your definition of namespace is that everything within the namespace has uniqueID references the same CAS/ObjectStore?

If that is the case, does this address the original concern that ActionCache key can just be a hash that doesn't anything in the object store? If that is possible, then what is the namespace of ActionCache (whose key and value are in different namespace)?

Reply:
The object store and action cache are fundamentally different. One is a set and the other is a map.

An **ObjectStore** is a set of objects. It has a schema for creating a content-derived unique identifier/handle for arbitrary objects, which can be used as a handle for retrieving on object later. This identifier must be stable, not dependent on state of the store itself, and uniquely identifying, not having collisions. So, a strong hash. While it is somewhat convenient to think of it as a map from a hash to an object, that's not really what's going on... it's really a set of objects, and objects have an inherent UID defined by their content.

This UID is universal (it's a UUID), assuming the universe is a particular CAS schema.

For example, the SHA1 identifiers used by Git are UUIDs *in the context of Git*. You can have multiple Git repos, each that have never talked to each other, and if you ingest the same filesystem directory you'll get the same UID for the tree object. But ingest that same directory into, say, a Bazel instance, and you'll get a different UID. This is again a UUID, but only in the context of *Bazel*.

You cannot take UIDs from one namespace and use them directly in another namespace.

An **ActionCache** is a map from an action (key) to a CAS object (value).
- At a high-level, an action (key) is just an arbitrary string. A particular action cache need not even store the string; maybe it'll use a strong hash algorithm and just store the hash. Different action cache implementations may make different choices here; could be a key-value store from the serialized action, or maybe from the hash of the serialized action. The **ActionDescription** provides a framework for serializing a complex action into a string (or directly into a hash).
- The CAS object (value) is not stored in the cache. Instead, only its UID is stored there.
- The ActionCache needs to know the namespace of the CAS objects it stores. Is this mapping to UIDs for a Git CAS, a Bazel CAS, or an LLVM builtin CAS?

> I guess I wasn't clear what is covered by a Namespace. Is that Key, or Value, or the mapping between Key and Value?

A **Namespace** defines the "universe" for UIDs. It answers the question for a UID: in what context are these universal / UUIDs? At a high-level, there is one namespace for each schema that converts from object content to its handle/UID. There would be one namespace referenced by all `ObjectStore`s wrapping a Git CAS, another for all `ObjectStore`s wrapping a Bazel CAS, and another for all `ObjectStore`s wrapping a builtin CAS. If we end up supporting alternative hash functions or UID lengths for the builtin CAS, then there will need to be a different namespace for each hash function.

> The Value/Result from ActionCache should be available as Key in ObjectStore but is very different from Value/Result in ObjectStore. I guess your definition of namespace is that everything within the namespace has uniqueID references the same CAS/ObjectStore?

No. You can have lots of ObjectStores (instances of a CAS) that all use the same namespace.

> If that is the case, does this address the original concern that ActionCache key can just be a hash that doesn't anything in the object store?

Yes, that's part of it.

> If that is possible, then what is the namespace of ActionCache (whose key and value are in different namespace)?

An ActionCache maps from arbitrary string to CAS objects. But it stores UIDs from the CAS objects' namespace. The ActionCache's namespace is the namespace for those CAS objects.

In theory, we could allow an ActionCache to store CAS objects from multiple namespaces. E.g., map "action1" to a Git object, and "action2" to a Bazel object. But I think this would be hard to use. Every time a client looked up a cached result, they'd need to check if the result was in the same namespace as the ObjectStore they have available before looking up the object.

Instead, the current design requires that an action cache only operates on CAS objects in a single namespace. Most parts of the compiler can reasonably assume that if they're handed an ObjectStore and an ActionCache, they're in the same namespace. They can verify this by asserting that `getNamespace()` returns the same thing... but there's no need to check namespaces after each `get()`.